### ISSUE TITLE
GraphBuilder API?

### ISSUE THREAD

**Issue by jeromekelleher** [popsim-consortium/demes-python#208]
This is going to be unpopular, so apologies in advance. 

I'm not a huge fan of the current API for building graphs incrementally in Python. I find the tension between being able to specify a graph concisely without all the unnecessary attributes and the desire to have a fully featured in-memory model of a Graph that you can be sure is non-redundant, etc, confusing.

Currently, we have things like
```python
    graph = demes.Graph(description="Island model", time_units="generations")
    for j, start_size in enumerate(size):
        graph.deme(
            id=f"pop_{j}", epochs=[demes.Epoch(start_size=start_size, end_time=0)]
        )
    graph.symmetric_migration(
        demes=[deme.id for deme in graph.demes], rate=migration_rate
    )
```

What if we had, instead,
```python
    gb = demes.GraphBuilder(description="Island model", time_units="generations")
    for j, start_size in enumerate(size):
        deme = gb.add_deme(id=f"pop_{j}")
        deme.add_epoch(start_size=start_size, end_time=0)
    gb.add_symmetric_migration(
        demes=[deme.id for deme in gb.demes], rate=migration_rate
    )
    graph = gb.finalise() # Or gb.bake(), or gb.resolve_all_the_things() or whatever
```

So, then the graph is considered to be fully read-only, and is guaranteed to have all the necessary attributes set. If you want to update an existing graph, you call ``gb = graph.builder()`` or something.

I think that ultimately the graphs will be fundamentally more useful if they are immutable; making code that reasons with them will be much easier in the long run. It also makes a natural point at which to run the graph resolution process, which we need for parsing. I don't see this currently with the current API?

Thoughts?

**Comment by grahamgower** [comment]
I considered a factory-style class (separate to the underlying Graph, Deme, Epoch, etc. objects) when first sketching things out, but I think I discarded the idea largely because adding epochs incrementally started getting awkward. In your example above, what would `GraphBuilder.add_deme()` return? A `Deme`, complete with `add_epoch()` method, or a `DemeBuilder` class? It would probably be simpler if add_epoch() was instead a method of GraphBuilder (e.g. `def GraphBuilder.add_epoch(*, deme, start_size, end_time, ...):`.

Given we now have a strong notion of a data model, we could have a GraphBuilder factory that operates on a data dictionary and then pass that into `Graph.fromdict()`. But then, we could do away with a class heirarchy altogether and always work with data dictionaries. Functions operating on data structures always seemed more natural to me than classes with methods anyhow---you never need to choose which subclass carries the method, because none of them do.

**Comment by jeromekelleher** [comment]
It is definitely useful to have a an in-memory model of the graph as classes: ``graph.demes[0].epochs[0]`` is much better than ``graph["demes"][0]["epochs"][0]``. This gets really tedious, *really* quickly (I have first-hand experience here from the reference implementation). So I'm a big fan of the simple object model built from data classes/attrs; this will be super useful over time as we build things that *do* things with demes graphs, other than *build* them (which is all we do at the moment, more or less). But, I think we're complicating this object model quite a lot currently with the complex logic needed for incrementally building, while still trying to maintain integrity of the data model.

>  In your example above, what would GraphBuilder.add_deme() return? A Deme, complete with add_epoch() method, or a DemeBuilder class? It would probably be simpler if add_epoch() was instead a method of GraphBuilder (e.g. def GraphBuilder.add_epoch(*, deme, start_size, end_time, ...):.

I guess probably a DemeBuilder class, if we're going to be fully consistent that the data model classes should be immutable. We could add the ``GraphBuilder.add_epoch`` method as you say, but this means that you need to type in the ID of the deme many times, rather than using a reference to it. That's not such a bit deal though, I think? At least we get to separate out the "what is a valid input" logic from "what is the structure of the data model".

**Comment by grahamgower** [comment]
I'm with you. But I definitely think this would be easiest if we just have the builder functions that operate on a data dict. E.g.

```python
class GraphFactory:
  def __init__(description, time_units, ...):
    self.data = dict(description=description, time_units=time_units, ...}

  def add_deme(id, ...):
    d = dict(id=id, ...)
    self.data["demes"].append(d)
    self.deme_dict[id] = d

  def add_epoch(deme_id, start_size, ...):
    e = dict(start_size=start_size, ...)
    self.deme_dict[deme_id]["epochs"].append(e)

  def resolve():
    return Graph.fromdict(self.data)
```

**Comment by jeromekelleher** [comment]
Yeah, that works. So the GraphFactory is about building up the data model from the spec, and the Graph then is the fully resolved object model, with extra bells and whistles. I like it.

**Comment by apragsdale** [comment]
I'm not sure if I follow perfectly the discussion around data structures/dictionaries/classes here.. I like the idea of having a `g.finalize()` or `g.resolve()` or similar command, and that `g` can be changed, updated, reconfigured using some API. I think it's important to keep in mind the developers (like with dadi and moments) who will want to be able to write functions that handle parameterized graphs and update graphs and parameters, such as epoch sizes and split times, change.

From what I can tell, nothing proposed here would impede that, is that right? I think the `GraphFactory` idea would get us closer to that perhaps, and I could even imagine having functions that `pop` or reattach demes or epochs or migrations, and consistency is only checked at the `resolve` stage. I'm just a bit worried about `demes` getting so narrow so that inference over a graph and its parameterization, even if developed in `moments` instead of `demes` itself, would not be possible.

I don't think my comment is very useful as far as weighing in data structures, etc. Just wanted to make sure we're thinking about possible outside developer applications and not constrain what's possible.

**Comment by grahamgower** [comment]
Separating the builder API from the resolved Graph ought to provide no less flexibility than we currently have. But it might make the flexibility more obvious to the user. I would hope that we make the `GraphBuilder.data` dict attribute part of the API and allow users to modify it directly. If the user makes changes to the `data` dict that can be successfully `resolve()`d, then the builder API should work fine after those user modifications.

Oh, and I don't have even a mild opinion on what it gets called: GraphBuilder, GraphFactory, GraphFrobber, ... meh.

**Comment by jeromekelleher** [comment]
I think the best way of thinking about it is this @apragsdale - the fully resolved data model is highly redundant, so it's hard to keep everything in sync.There's a tension between making a fully resolved object model which contains all the things someone might want like ``Deme.end_time`` (or whatever) and providing a robust way of building graphs from the minimal data model (which is what we should encourage people to do, so that everything is well defined).

So, we should **describe** graphs using the minimal model in the spec, but **work with** graphs using the fully resolved and augmented graph model like we currently have. The proposal here is to split the API into two, so that we have different APIs for the different tasks.

**Comment by apragsdale** [comment]
Ok that all makes sense to me. This does sound like a pretty major overhaul that is being proposed - @grahamgower do you have a good idea of how you envision it working? I wonder if we should have a quick call at some point to organize about all this a bit more, and set some goals for our 0.x release. I imagine that we'll want to have this 0.1 release get out there, let people play with it, and expect to have to make some minor changes before a subsequent more cemented 1.0.

**Comment by grahamgower** [comment]
> Ok that all makes sense to me. This does sound like a pretty major overhaul that is being proposed - @grahamgower do you have a good idea of how you envision it working?

Yeah, I think so. It won't be too complicated to make the changes in `demes/demes.py`, but the corresponding changes throughout the tests will probably be a lot of messy churn.

> I wonder if we should have a quick call at some point to organize about all this a bit more, and set some goals for our 0.x release.

Sure. I was hoping we only had a couple of things left to knock off before having another prerelease for more widespread consumption. But this API change would probably delay that.

> I imagine that we'll want to have this 0.1 release get out there, let people play with it, and expect to have to make some minor changes before a subsequent more cemented 1.0.

Yep, that's exactly what I was thinking.

**Comment by grahamgower** [comment]
This was a great idea @jeromekelleher! It simplified lots of things! Closed in #209.
